testsuite/tests/quotation/quotation_typing.ml

```diff
@@ -0,0 +1,254 @@
+|}];;
+
+type t4 = A | B;;
+[%%expect {|
+type t4 = A | B
+|}];;
+
+<[A]>;;
+[%%expect {|
+Line 1, characters 2-3:
+1 | <[A]>;;
+      ^
+Error: Constructor "A" used at Line 1, characters 2-3
+       is unbound in this context; identifier "A" is unbound
+       in a context with one layer of quotation (<[ ... ]>).
+|}];;
+
+<[fun (x : 'a) (y : 'b) -> (x, y)]>;;
+[%%expect {|
+- : <[ $ ('a) -> $ ('b) -> $ ('a) * $ ('b) ]> expr =
+<[ fun (x : 'a) (y : 'b) -> (x, y) ]>
+|}];;
+
+<[fun (f : 'a. 'a -> 'a) (x : 'b) -> f x]>;;
+[%%expect {|
+- : <[ ('a. 'a -> 'a) -> $ ('b) -> $ ('b) ]> expr =
+<[ fun (f : 'a. 'a -> 'a) (x : 'b) -> f x ]>
+|}];;
+
+<[fun (f : 'a. 'a -> 'a) (x : 'a) -> f x]>;;
+[%%expect {|
+- : <[ ('a. 'a -> 'a) -> $ ('a) -> $ ('a) ]> expr =
+<[ fun (f : 'a. 'a -> 'a) (x : 'a__1) -> f x ]>
+|}];;
+
+<[fun (x : 'a) (f : 'a. 'a -> 'a) -> f x]>;;
+[%%expect {|
+- : <[ $ ('a) -> ('a0. 'a0 -> 'a0) -> $ ('a) ]> expr =
+<[ fun (x : 'a) (f : 'a__1. 'a__1 -> 'a__1) -> f x ]>
+|}];;
+
+<[fun (f : 'a. 'a -> 'a) (g: 'b 'c. 'b list -> ('b -> 'c) -> 'c list) -> f g]>;;
+[%%expect {|
+- : <[
+     ('a. 'a -> 'a) ->
+     ('b 'c. 'b list -> ('b -> 'c) -> 'c list) ->
+     $ ('d) list -> ($ ('d) -> $ ('e)) -> $ ('e) list ]>
+    expr
+=
+<[
+  fun (f : 'a. 'a -> 'a) (g : 'b 'c. 'b list -> ('b -> 'c) -> 'c list) -> f g
+]>
+|}];;
+
+(* The mk_pair examples exist to test whether unification behaves well when
+   splices and quotations are present. *)
+
+let mk_pair x = <[$x, $x]>;;
+[%%expect {|
+val mk_pair : <[ $ ('a) ]> expr -> <[ $ ('a) * $ ('a) ]> expr = <fun>
+|}];;
+
+mk_pair <[123]>;;
+[%%expect {|
+- : <[ $ (<[ int ]>) * $ (<[ int ]>) ]> expr = <[ (123, 123) ]>
+|}];;
+
+mk_pair <[[]]>;;
+[%%expect {|
+- : <[ $ (<[ $ ('a) list ]>) * $ (<[ $ ('a) list ]>) ]> expr = <[ ([], []) ]>
+|}];;
+
+mk_pair <[None]>;;
+[%%expect {|
+- : <[ $ (<[ $ ('a) option ]>) * $ (<[ $ ('a) option ]>) ]> expr =
+<[ (None, None) ]>
+|}];;
+
+mk_pair <[Some 123]>;;
+[%%expect {|
+- : <[ $ (<[ int option ]>) * $ (<[ int option ]>) ]> expr =
+<[ ((Some 123), (Some 123)) ]>
+|}];;
+
+mk_pair <[fun () -> 42]>;;
+[%%expect {|
+- : <[ $ (<[ unit -> int ]>) * $ (<[ unit -> int ]>) ]> expr =
+<[ ((fun () -> 42), (fun () -> 42)) ]>
+|}];;
+
+mk_pair <[fun x -> x]>;;
+[%%expect {|
+- : <[ $ (<[ '_weak1 -> '_weak1 ]>) * $ (<[ '_weak1 -> '_weak1 ]>) ]> expr =
+<[ ((fun x -> x), (fun x -> x)) ]>
+|}];;
+
+(* Type algebra checks. *)
+
+fun (x: 'a) -> (x: <[<[<[$($($'a))]>]>]>);;
+[%%expect {|
+- : 'a -> <[ <[ <[ $ ($ ($ ('a))) ]> ]> ]> = <fun>
+|}];;
+
+fun (x: <[<[<[$($($'a))]>]>]>) -> (x: 'a);;
+[%%expect {|
+- : <[ <[ <[ $ ($ ($ ('a))) ]> ]> ]> -> 'a = <fun>
+|}];;
+
+fun (x: <[<[<[$($'a)]>]>]>) -> (x: 'a);;
+[%%expect {|
+Line 1, characters 32-33:
+1 | fun (x: <[<[<[$($'a)]>]>]>) -> (x: 'a);;
+                                    ^
+Error: This expression has type "<[ 'a ]>"
+       but an expression was expected of type "'a"
+       The type variable "'a" occurs inside "<[ 'a ]>"
```

Comment:
I don't think you should be able to get to an occurs check like this: there should already have been an error complaining that a variable was occuring at multiple stages.